### PR TITLE
Switching to using PHP-FPM

### PR DESCRIPTION
**Pull request by oparoz**
@kyrofa @jospoortvliet @ezraholm50 @enoch85 @miska

Snaps of this PR for your beta PiClouds are available from here:
https://code.launchpad.net/~olivier-interfasys/+snap/owncloud-snap

Sideloading instructions:
https://github.com/owncloud/pi-image/wiki/Sideloading-snaps-for-testing

I've added comments in-line to explain the various components.
### TODO
- [x] Remove `export OWNCLOUD_DATABASE_PASSWORD=$(cat $owncloud_password_path)`
- [x] Add PHP extensions to PHP cli. Can be another PR

### CODE REVIEW

File: parts/plugins/x-apache.py [hunk 1]
```diff
@@ -185,7 +185,7 @@ def build(self):
 
         shutil.copytree(apache_source_directory, apache_build_directory)
 
-        subprocess.check_call("./configure --prefix={} --enable-modules=none --enable-mods-shared='{}' ENABLED_DSO_MODULES='{}'".format(self.installdir, ' '.join(self.options.modules), ','.join(self.options.modules)),
+        subprocess.check_call("./configure --prefix={} --with-mpm=event --enable-modules=none --enable-mods-shared='{}' ENABLED_DSO_MODULES='{}'".format(self.installdir, ' '.join(self.options.modules), ','.join(self.options.modules)),
```
Comment (oparoz) [inline parts/plugins/x-apache.py:186]: Force event MPM to lower memory usage.

File: parts/plugins/x-apache.py [hunk 2]
```diff
@@ -247,7 +247,6 @@ def _build_third_party_modules(self):
 
             configure_command = [
                 './configure', '--prefix=' + self.installdir,
-                '--with-apxs2={}/bin/apxs'.format(self.installdir),
```
Comment (oparoz) [inline parts/plugins/x-apache.py:248]: Remove mod_php

File: snapcraft.yaml [hunk 3]
```diff
@@ -1,8 +1,7 @@
 name: owncloud
 version: 9.0.0ubuntu1
 summary: ownCloud
-description: ownCloud running on Apache with MySQL. This is currently in beta.
-icon: icon.svg
```
Comment (oparoz) [inline snapcraft.yaml]: The icon currently gets in the way, will probably be restored before merging

File: snapcraft.yaml [hunk 4]
```diff
@@ -12,6 +11,13 @@ apps:
     daemon: simple
     plugs: [listener]
 
+  # PHP-FPM daemon
```
Comment (oparoz) [inline snapcraft.yaml:14]: That's the daemon responsible for keeping PHP-FPM alive and ready to process requests

File: snapcraft.yaml [hunk 5]
```diff
@@ -45,7 +51,8 @@ parts:
     # The built-in Apache modules to enable
     modules:
       - headers
-      - fcgid
+      - proxy
+      - proxy_fcgi
```
Comment (oparoz) [inline snapcraft.yaml:67]: We use mod_proxy_fcgi now

File: snapcraft.yaml [hunk 6]
```diff
@@ -63,20 +70,29 @@ parts:
...
     third-party-modules:
       # Build PHP
-      - source: http://us1.php.net/get/php-7.0.2.tar.gz/from/this/mirror
+      - source: http://us1.php.net/get/php-7.0.5.tar.gz/from/this/mirror
         source-type: tar
         configflags:
+          - --enable-fpm
+          - --disable-pear
+          - --disable-phar
+          - --disable-phpdbg
```
Comment (oparoz) [inline snapcraft.yaml]: Enabling PHP-FPM, disabling unneeded things.
People won't be installing libs via PEAR or launch PHAR executables
We use xdebug and not phpdbg

File: snapcraft.yaml [hunk 7]
```diff
@@ -63,20 +70,29 @@ parts:
...
           - --enable-ctype
           - --enable-mbstring
           - --enable-zip
           - --with-pdo-mysql
           - --with-zlib
           - --with-gd
+          - --enable-gd-native-ttf
+          - --with-jpeg-dir=/usr/lib
+          - --with-freetype-dir=/usr/lib
```
Comment (oparoz) [inline snapcraft.yaml]: That's issue #26 already fixed. Can be easily removed

File: snapcraft.yaml [hunk 8]
```diff
@@ -63,20 +70,29 @@ parts:
...
           - --with-curl
           - --with-openssl
           - --with-bz2
           - --with-mcrypt
           - --enable-exif
+          - --enable-pcntl
+          - --enable-intl
```
Comment (oparoz) [inline snapcraft.yaml]: Some missing extensions, more to come in other PRs

File: snapcraft.yaml [hunk 9]
```diff
@@ -88,18 +104,25 @@ parts:
...
 
   # Copy over our PHP configuration file.
   php-config:
     plugin: copy
     files:
-      src/php/php.ini: php.ini
+      src/php/php.ini: php/lib/php.ini
+      src/php/extensions.ini: php/lib/php.conf.d/extensions.ini
+      src/php/php-fpm.conf: php/etc/php-fpm.conf
+      src/php/php-fpm: bin/php-fpm
```
Comment (oparoz) [inline snapcraft.yaml]: New config files and the init script for php-fpm

File: snapcraft.yaml (hunk 7 above)
Comment (kyrofa) [inline snapcraft.yaml]: Though that fix still needs to be proposed.

File: src/owncloud/apache_config [hunk 10]
```diff
@@ -1,9 +1,10 @@
-<FilesMatch \.php$>
-    SetHandler application/x-httpd-php
+# PHP-FPM
+SetEnvIfNoCase ^Authorization$ "(.+)" HTTP_AUTHORIZATION=$1
+ProxyTimeout 900
+<FilesMatch "\.(inc|php)$">
+	SetHandler "proxy:unix:${SNAP_DATA}/php/sockets/www-data.sock|fcgi://localhost"
```
Comment (oparoz) [inline src/owncloud/apache_config]: Using mod_proxy_fcgi to connect Apache and PHP-FPM

File: src/owncloud/apache_config [hunk 11]
```diff
@@ -14,3 +15,14 @@ Alias "/apps" "${SNAP_DATA}/owncloud/apps"
...
+# MPM Event
+<IfModule event.c>
+	StartServers			2
+	MinSpareThreads			4
+	MaxSpareThreads			32
+	ThreadsPerChild			16
+	ServerLimit				4
+	MaxRequestWorkers		64
+	MaxConnectionsPerChild	0
+</IfModule>
```
Comment (oparoz) [inline src/owncloud/apache_config]: An example mpm_event config, can easily be modified here or in another PR

File: src/owncloud/autoconfig.php [hunk 12]
```diff
@@ -3,23 +3,26 @@
...
 
 $snap_fullname = $snap_name . '.' . $snap_developer;
 
-$data_path = '/var/lib/snaps/'.$snap_fullname.'/current';
+$data_path = '/var/lib/snaps/' . $snap_fullname . '/current';
 
-$AUTOCONFIG = array(
-'directory' => $data_path.'/owncloud/data',
+$writable = getenv('SNAP_DATA');
+$owncloud_password = trim(file_get_contents($writable . '/mysql/owncloud_password'));
```
Comment (oparoz) [inline src/owncloud/autoconfig.php:16]: Fetching the DB password from the file directly instead of using environment variables

File: src/php/extensions.ini [hunk 13]
```diff
@@ -0,0 +1,5 @@
+; Extensions are located in $SNAP/lib/php/extensions/no-debug-non-zts-20151012
+; OpCache
+zend_extension=${SNAP}/lib/php/extensions/no-debug-non-zts-20151012/opcache.so
```
Comment (oparoz) [inline src/php/extensions.ini:3]: Loading opcache, may require some tuning, see #31

File: src/php/php-fpm [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+
+# Setup PHP
+PHPFOLDER=$SNAP_DATA/php
+if [ ! -d "$PHPFOLDER" ]; then
+	echo "Configuring PHP-FPM..."
+	mkdir -p $PHPFOLDER
+	mkdir -p $PHPFOLDER/log
+	touch $PHPFOLDER/log/php-fpm.log
+	mkdir -p $PHPFOLDER/sockets
+fi
```
Comment (oparoz) [inline src/php/php-fpm]: That setup routine could be placed in a separate script, if better in terms of management.

File: src/php/php-fpm.conf [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+; Default Value: notice
+log_level = notice
+
+;;;;;;;;;;;;;;;;;;;;
+; Pool Definitions ;
+;;;;;;;;;;;;;;;;;;;;
+
+[www-data]
+user = root
+group = root
```
Comment (oparoz) [inline src/php/php-fpm.conf:15]: We have to use root!

File: src/php/php-fpm [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+# config: $SNAP_DATA/php/php-fpm.conf
+
+PID=$SNAP_DATA/php/php-fpm.pid
+BIN=$SNAP/sbin/php-fpm
+CONFFOLDER=$SNAP
+FPMCONF=$CONFFOLDER/php/etc/php-fpm.conf
+CONFINI=$CONFFOLDER/php/lib/php.ini
+PHP_INI_SCAN_DIR=$SNAP/php/lib/php.conf.d
+export PHP_INI_SCAN_DIR
+OPTS="-R --daemonize --pid=${PID} --fpm-config $FPMCONF -c $CONFINI"
```
Comment (oparoz) [inline src/php/php-fpm]: `-R` allows us to run PHP-FPM as root.

File: src/php/php-fpm.conf [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+listen = ${SNAP_DATA}/php/sockets/$pool.sock
+; Any process using setuid is instantly killed
+;listen.owner = $pool
+;listen.group = $pool
+;listen.mode = 660
+
+pm = ondemand
+pm.max_children = 9
+pm.process_idle_timeout = 60
+pm.max_requests = 200
```
Comment (oparoz) [inline src/php/php-fpm.conf]: Can be tuned, see #28

File: src/php/php.ini [hunk 18]
```diff
@@ -186,7 +186,7 @@
 
 ; Enable the PHP scripting language engine under Apache.
 ; http://php.net/engine
-engine = On
+engine = Off
```
Comment (oparoz) [inline src/php/php.ini:189]: Disable mod_php

File: src/php/php.ini [hunk 19]
```diff
@@ -720,7 +720,7 @@ user_dir =
 
 ; Directory in which the loadable extensions (modules) reside.
 ; http://php.net/extension-dir
-; extension_dir = "./"
+extension_dir = "${SNAP}/lib/php/extensions/no-debug-non-zts-20151012"
```
Comment (oparoz) [inline src/php/php.ini:723]: Set the extension folder so that PHP-FPM can find our future extensions.

**Comment by oparoz** [comment]
Snaps of this PR for your beta PiClouds are available from here:
https://code.launchpad.net/~olivier-interfasys/+snap/owncloud-snap

**Comment by oparoz** [comment]
It makes a huge difference on my device. 
If someone could look at theirs and give me the amount of RAM they have free when syncing.
This is what I have now:

```
$ vmstat 5
procs -----------memory---------- ---swap-- -----io---- -system-- ------cpu-----
 r  b   swpd   free   buff  cache   si   so    bi    bo   in   cs us sy id wa st
 3  0      0 335956  75304 305636    0    0     0  1823 4691 7146 50 20 24  6  0
```

**Comment by oparoz** [comment]
Sideloading instructions:
https://github.com/owncloud/pi-image/wiki/Sideloading-snaps-for-testing

**Comment by jospoortvliet** [comment]
Oh halleluja, great work Oparoz. I will give this a test tomorrow, won't have time today. But rockingly cool!

File: snapcraft.yaml (hunk 3 above)
Comment (kyrofa) [inline snapcraft.yaml]: The bug that you were hitting has been fixed in the most recent release of Snapcraft (2.7). Please update when you're able and put this back.

File: parts/plugins/x-apache.py (hunk 2 above)
Comment (kyrofa) [inline parts/plugins/x-apache.py:248]: Yeah you should probably not cannibalize the apache plugin here-- FPM doesn't depend on apache, so it doesn't belong here.

File: src/owncloud/autoconfig.php (hunk 12 above)
Comment (kyrofa) [inline src/owncloud/autoconfig.php:16]: Why? Particularly since you didn't remove the variable itself?

File: snapcraft.yaml (hunk 9 above)
Comment (kyrofa) [inline snapcraft.yaml]: As of Snapcraft v2.7 you can leave off the destination file names if you don't want to rename them, so this becomes:

``` yaml
src/php/php.ini: php/lib/
src/php/extensions.ini: php/lib/php.conf.d/
# ...
```

Less typing, less error-prone.

File: src/owncloud/apache_config (hunk 11 above)
Comment (kyrofa) [inline src/owncloud/apache_config]: Yeah let's remove this here. This rpi-specific stuff will probably need to go in the fork.

File: src/php/php-fpm [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+
+#ulimit -HSn 32768
```
Comment (kyrofa) [inline src/php/php-fpm]: No need to include crufty comments.

File: snapcraft.yaml (hunk 3 above)
Comment (oparoz) [inline snapcraft.yaml]: Thanks

File: snapcraft.yaml (hunk 9 above)
Comment (oparoz) [inline snapcraft.yaml]: That's good news :)

File: parts/plugins/x-apache.py (hunk 2 above)
Comment (oparoz) [inline parts/plugins/x-apache.py:248]: Yeah, will give it a go

File: src/php/php-fpm (hunk 20 above)
Comment (oparoz) [inline src/php/php-fpm]: How do we set that ulimit though?

File: src/owncloud/apache_config (hunk 11 above)
Comment (oparoz) [inline src/owncloud/apache_config]: Actually, I had forgotten that this only works with PHP-FPM, so it's best to keep a sane starting point here, which can be tuned later, depending on when and if we agree on a better config.

File: src/owncloud/autoconfig.php (hunk 12 above)
Comment (oparoz) [inline src/owncloud/autoconfig.php:16]: Removing the variable is already on the todo, that env variable is not always available to PHP-FPM and can't be added afaik.

Comment (kyrofa) [inline src/owncloud/autoconfig.php:16]: Ah right, because now it's a separate process. Keep in mind as you refactor this that (especially on the pi) mysql takes quite some time to set itself up and generate the password upon first install. You don't want to evaluate this config file until that file exists.

File: src/php/php-fpm (hunk 20 above)
Comment (kyrofa) [inline src/php/php-fpm]: I doubt you can from within the snap. You typically can't affect anything system-wide.

File: src/owncloud/apache_config (hunk 11 above)
Comment (kyrofa) [inline src/owncloud/apache_config]: Just so I understand, are you saying that php-fpm _requires_ this?

File: src/php/php.ini (hunk 19 above)
Comment (kyrofa) [inline src/php/php.ini:723]: `no-debug-non-zts-20151012` feels like a build artifact. Is it possible to use `${SNAP}/lib/php/extensions/` ?

File: src/owncloud/apache_config (hunk 11 above)
Comment (oparoz) [inline src/owncloud/apache_config]: It's not a hard requirement, but they go hand in hand. You don't get all the benefits of using PH-FPM without using mpm_event.

File: src/owncloud/autoconfig.php (hunk 12 above)
Comment (oparoz) [inline src/owncloud/autoconfig.php:16]: From what I understand, Apache won't start before this is set up, so the user won't be able to open the page before that password is written to disk.

Comment (kyrofa) [inline src/owncloud/autoconfig.php:16]: Right, but if you remove the fact that the variable is defined there it's really waiting for a file to exist that it doesn't use. I wonder if there's a better place for it.

File: src/php/php.ini (hunk 19 above)
Comment (oparoz) [inline src/php/php.ini:723]: It's actually a standard path, sort of like an ABI identifier.

Comment (kyrofa) [inline src/php/php.ini:723]: Ah, okay :smiley: .

File: src/owncloud/autoconfig.php (hunk 12 above)
Comment (oparoz) [inline src/owncloud/autoconfig.php:16]: > it's really waiting for a file to exist that it doesn't use.

I don't understand. It does use the file

Comment (kyrofa) [inline src/owncloud/autoconfig.php:16]: > I don't understand. It does use the file

It does now, but only to define the variable. We'll just need to make sure we document what we're doing that so it doesn't appear crufty is all :smiley: . The only other place to put it is the php-fpm startup script, and I don't really feel like it belongs there either.

Comment (oparoz) [inline src/owncloud/autoconfig.php:16]: > It does now, but only to define the variable

I still don't get it :D. The variable is used when setting things up, then the file is destroyed and the variable doesn't exist any more anywhere.

> We'll just need to make sure we document what we're doing that so it doesn't appear crufty is all :smiley:

I'll add a comment.

> The only other place to put it is the php-fpm startup script, and I don't really feel like it belongs there either.

No it doesn't belong there. That file is loaded every time a PHP-FPM process is started.

Comment (kyrofa) [inline src/owncloud/autoconfig.php:16]: > I still don't get it :D. The variable is used when setting things up, then the file is destroyed and the variable doesn't exist any more anywhere.

I'm talking about [this](https://github.com/kyrofa/owncloud-snap/blob/master/src/owncloud/setup_owncloud#L12). All it uses the file for is to define that variable. If you remove the variable definition from it, then all you have left is "I'm waiting for this file.... okay move on." Though honestly that's no different than what we do for MySQL only a few lines up. The file isn't actually destroyed, and the variable is defined every time (so I quite like your change). I don't think the file should be destroyed, but the variable is definitely not necessary anymore.

Comment (oparoz) [inline src/owncloud/autoconfig.php:16]: Ah, it's clearer now, so I don't see a problem with having to wait for the password to be defined and saved to file in that `setup_owncloud` script. We have to wait for the DB to be set up anyway.

> The file isn't actually destroyed

I was talking about autoconfig.php

> but the variable is definitely not necessary anymore.

It's already remove in what I'll push later

File: src/owncloud/apache_config (hunk 11 above)
Comment (kyrofa) [inline src/owncloud/apache_config]: Okay. Just try to keep configs pretty general in this repo-- we can really tune it for the rpi in the fork. I'll trust you here, since I know nothing about these values :wink: .

**Comment by kyrofa** [comment]
After you've finished refactoring this, please make the PR into develop instead of master (more information in the [contribution guide](https://github.com/kyrofa/owncloud-snap/blob/master/CONTRIBUTING.md)).

**Comment by oparoz** [comment]
:+1:

**Comment by oparoz** [comment]
Replaced by #52
